File: app/models/sensor.py
from sqlalchemy import Column, DateTime, ForeignKey, String, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.core.database import Base


class Sensor(Base):
    __tablename__ = "sensors"

    id = Column(UUID(as_uuid=True), primary_key=True, server_default=func.gen_random_uuid())
    asset_id = Column(
        UUID(as_uuid=True),
        ForeignKey("assets.id", ondelete="CASCADE"),
        nullable=False,
    )
    name = Column(String(255), nullable=False)
    sensor_type = Column(String(100), nullable=False)
    unit = Column(String(50))
    status = Column(String(50), nullable=False, default="active")
    created_at = Column(DateTime(timezone=True), nullable=False, server_default=func.now())
    updated_at = Column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

    asset = relationship("Asset", back_populates="sensors")
    readings = relationship("SensorData", back_populates="sensor", cascade="all, delete-orphan")

    def to_dict(self) -> dict:
        return {
            "id": str(self.id),
            "asset_id": str(self.asset_id),
            "name": self.name,
            "sensor_type": self.sensor_type,
            "unit": self.unit,
            "status": self.status,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }

File: frontend/src/types/sensor.ts
export type SensorStatus = "active" | "inactive" | "faulty";

export interface Sensor {
  id: string;
  asset_id: string;
  name: string;
  sensor_type: string;
  unit: string | null;
  status: SensorStatus;
  created_at: string | null;
  updated_at: string | null;
}

export interface SensorCreate {
  name: string;
  sensor_type: string;
  unit?: string;
  status?: SensorStatus;
}

export interface SensorUpdate {
  name?: string;
  sensor_type?: string;
  unit?: string;
  status?: SensorStatus;
}
